File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    pair<int,int> checkPalindrome(string& s,int l,int r){
        int prevl=-1,prevr=-1;
        while(l>=0 && r<s.length() && s[l]==s[r]){
            prevl=l;
            prevr=r;
            l--;r++;
        }
        return {prevl,prevr};
    }
    string longestPalindrome(string s) {
        // aba
        // abba
        int n = s.size(); 
        int res = 0,l=-1,r=-1;
        for(int i=0;i<n;i++)
        {
            auto [l1,r1]=checkPalindrome(s,i,i);
            auto [l2,r2]= checkPalindrome(s,i,i+1);
            int len1 = (r1-l1)+1;
            int len2 = (r2-l2)+1;
            if(len1>res){
                res=len1;
                l=l1;r=r1;
            }
            if(len2>res){
                res=len2;
                l=l2,r=r2;
            }
        } 
        return s.substr(l,res);
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def checkPal(self,s,l,r): 
        while l>=0 and r<len(s) and s[l]==s[r]:
            l-=1 
            r+=1 
        r-=1 
        l+=1 
        return r-l+1,l,r
    def longestPalindrome(self, s: str) -> str:
        n = len(s) 
        res =""
        for i in range(n): 
            # for even i,i+1 
            # for odd i,i
            a,al,ar = self.checkPal(s,i,i) 
            b,bl,br = self.checkPal(s,i,i+1)
            if a>len(res): 
                res = s[al:ar+1]
            if b>len(res): 
                res = s[bl:br+1]
        return res